api/turing/batch/runner/ensembling.go

```diff
@@ -0,0 +1,322 @@
+	ensemblingJob *models.EnsemblingJob,
+	mlpProject *mlp.Project,
+) {
+	state, err := r.ensemblingController.GetStatus(mlpProject.Name, ensemblingJob)
+	if err != nil {
+		log.Errorf("Unable to get status of spark application: %v", err)
+		return
+	}
+
+	if state == batchcontroller.SparkApplicationStateUnknown {
+		// Do nothing, just wait to see if something happens
+		log.Warnf("Spark application state is unknown: %s", ensemblingJob.Name)
+		return
+	}
+
+	if state == batchcontroller.SparkApplicationStateCompleted {
+		ensemblingJob.Status = models.JobCompleted
+	} else if state == batchcontroller.SparkApplicationStateFailed {
+		ensemblingJob.Status = models.JobFailed
+	}
+
+	err = r.ensemblingJobService.Save(ensemblingJob)
+	if err != nil {
+		log.Errorf("Unable to save ensemblingJob %d: %v", ensemblingJob.ID, err)
+		return
+	}
+}
+
+func (r *ensemblingJobRunner) processBuildingImage(
+	ensemblingJob *models.EnsemblingJob,
+	mlpProject *mlp.Project,
+) {
+	status, err := r.imageBuilder.GetImageBuildingJobStatus(
+		mlpProject.Name,
+		ensemblingJob.InfraConfig.EnsemblerName,
+		ensemblingJob.EnsemblerID,
+	)
+
+	if status == imagebuilder.JobStatusActive {
+		// Do nothing
+		return
+	}
+
+	// We retry on all other possible outcomes
+	ensemblingJob.Status = models.JobPending
+	ensemblingJob.IsLocked = false
+	ensemblingJob.RetryCount++
+	saveErr := r.ensemblingJobService.Save(ensemblingJob)
+	if saveErr != nil {
+		log.Errorf("Unable to save ensemblingJob %d: %v", ensemblingJob.ID, err)
+	}
+}
+
+func (r *ensemblingJobRunner) processJobs() {
+	isLocked := false
+	options := service.EnsemblingJobListOptions{
+		PaginationOptions: service.PaginationOptions{
+			Page:     testutils.NullableInt(1),
+			PageSize: &r.batchSize,
+		},
+		Statuses: []models.Status{
+			models.JobPending,
+			models.JobFailedSubmission,
+			models.JobFailedBuildImage,
+		},
+		IsLocked:           &isLocked,
+		RetryCountLessThan: &r.maxRetryCount,
+	}
+	ensemblingJobsPaginated, err := r.ensemblingJobService.List(options)
+	if err != nil {
+		log.Errorf("unable to query ensembling jobs: %v", err)
+	}
+
+	if ensemblingJobsPaginated.Results == nil {
+		return
+	}
+	ensemblingJobs := ensemblingJobsPaginated.Results.([]*models.EnsemblingJob)
+	for _, ensemblingJob := range ensemblingJobs {
+		// Don't bother waiting, let the person calling the
+		// BatchJobRunner interface decide how many to submit at once.
+		go r.processOneEnsemblingJob(ensemblingJob)
+	}
+}
+
+func (r *ensemblingJobRunner) unlockJob(ensemblingJob *models.EnsemblingJob) {
+	ensemblingJob.IsLocked = false
+	err := r.ensemblingJobService.Save(ensemblingJob)
+	if err != nil {
+		log.Errorf("unable to unlock ensembling job %v", err)
+	}
+}
+
+func (r *ensemblingJobRunner) processOneEnsemblingJob(ensemblingJob *models.EnsemblingJob) {
+	defer r.unlockJob(ensemblingJob)
+
+	// lock ensembling job so others parallel processes don't take it
+	ensemblingJob.IsLocked = true
```

Comment:
The batch processing is performed by a single go routine, right? It's not possible to have any race conditions on this lock?

Reply:
Multiple instances of the application might cause a race condition.

Reply:
> The batch processing is performed by a single go routine, right?

Actually this is not true. https://github.com/ashwinath/turing/blob/batch-ensembler-orchestration/api/turing/batch/runner/runner.go#L16 It fires all the time, so we don't want it picking locked records as well.

Reply:
Sorry, perhaps I didn't phrase my question clearly. The race condition I was referring to is in accessing the DB.

The runner fires all the time, starting a set of goroutines with each iteration. But currently, we have a single runner ([ref](https://github.com/ashwinath/turing/blob/batch-ensembler-orchestration/api/turing/api/appcontext.go#L134)) - so, in each iteration, there is only one goroutine started.

If we had multiple goroutines in each iteration, more than one of which will examine the same set of DB records, it is possible that more than one runner picks up the same record for processing and sets `IsLocked` to `true` (because querying the jobs is separate from setting IsLocked=true - they are not synchronised).

And even when we have a single runner, AFAIK, the [time.Sleep](https://github.com/ashwinath/turing/blob/batch-ensembler-orchestration/api/turing/batch/runner/runner.go#L18) doesn't necessarily guarantee that the goroutine from the previous iteration would have done its checks by the time the next one fires. It can be influenced by many factors like IO waits and other resources. Because they are goroutines, the order of execution is not guaranteed. Although, since the interval between each iteration is in the order of seconds, realistically, this shouldn't be an issue.

> The batch processing is performed by a single go routine, right?

By this, I meant, within a single iteration, we only have one runner routine. If we are sure that this will be the case going forward as well, it is not as concerning. But, in the failproof implementation, the reading of the states from the DB and its update should be done together, and no other goroutines should be allowed to do this exact operation concurrently. So we would need a Mutex lock: https://golang.org/pkg/sync/#Mutex.Lock

Does it make sense?

Reply:
Crystal, but some points I want to clarify regarding my intention of this design.

> By this, I meant, within a single iteration, we only have one runner routine. If we are sure that this will be the case going forward as well, it is not as concerning.

This is not what I want. I want the goroutines to keep firing non stop, with a small delay like 5-10 seconds. There is minimal processing required in this goroutine. While this might look like there will be too many goroutines, most of the time its waiting on condition (waiting on kubernetes job to be done or I/O events like calling kubernetes/MLP etc).

The job can be simplified to this:

1. processJob()
  * query records (point A, https://github.com/ashwinath/turing/blob/batch-ensembler-orchestration/api/turing/batch/runner/ensembling.go#L198)
  * lock records (point B, https://github.com/ashwinath/turing/blob/batch-ensembler-orchestration/api/turing/batch/runner/ensembling.go#L228)
  * create Job in Kubernetes to build image
  * wait for job to be done (sleeping)
  * create secrets/service accounts and whatsoever
  * create spark job (does not wait)
2. updateStatus()
  * poll image building to check if it has hanged
  * check spark job if its done (does not wait)
  
So ideally Point A and B should be locked and not the entire lifetime of the goroutine (Although I believe that we should have done something like `update record set is_locked=true ... where ... RETURNING ...` which Gorm does not allow; see https://github.com/go-gorm/gorm/issues/4088). This would guarantee that it will not double process a job. Even if it double processes, it shouldn't be too much of a problem as the job will get overriden, skip steps if the image has been built.

It's definitely correct that there might be a race condition here, but I think that the chances of it happening are extremely low if you were to put a delay like 5-10s here: https://github.com/ashwinath/turing/blob/batch-ensembler-orchestration/api/turing/batch/runner/runner.go#L18. Since we don't need absolute accuracy on this, eventually, this will be consistent since any amount of execution will eventually yield the same results. While this can be solved with using channels, the problems will still surface if we are doing something like multi API deployments (which from my understanding we aren't doing it now except during rolling updates).

Anyway, I'll try to fix this. There seems to be a solution using a raw query with gorm instead: https://github.com/go-gorm/gorm/issues/4433#issuecomment-854225876

Reply:
Actually I'm not sure if it's worth the fix, seems like it will break the pattern of using the ORM because I have to write the SQL. If it does become a problem then I'll fix it? Are you okay with this? or do u want me write something like this:
```go
db.Raw(
		"UPDATE users SET name = ?, updated_at = ? WHERE age = ? RETURNING id, name, age, updated_at",
		"User0",
		time.Now(),
		20,
	).Scan(&users)
```

Reply:
> So ideally Point A and B should be locked and not the entire lifetime of the goroutine 

I agree with this.

> While this can be solved with using channels, the problems will still surface if we are doing something like multi API deployments ...

* I believe passing around a pointer to a `sync.Mutex` to all similar runners will suffice, we don't need channels.
* Btw, what happens during rollings updates or otherwise, if the app acquires a DB lock but dies before submitting the Job to kubernetes? In this event, is it right to say that we will hit timeout and the job will be picked up again in the next iteration?

Reply:
> I think that the chances of it happening are extremely low if you were to put a delay like 5-10s here

I'm ok with this. Could you perhaps leave a comment in the code summarizing this? As you're saying, we can revisit it if there is an issue.

Reply:
> Btw, what happens during rollings updates or otherwise, if the app acquires a DB lock but dies before submitting the Job to kubernetes? In this event, is it right to say that we will hit timeout and the job will be picked up again in the next iteration?

Yes, and my bet is that since this is a batch process, time shouldn't be too sensitive here, roughly 10 minutes of delay should not be devastating.

Reply:
Yah I'll add in a comment on this.